akka-docs-dev/rst/scala/stream-customize.rst

```diff
@@ -9,8 +9,162 @@ is sometimes necessary to define new transformation stages either because some f
+* ``grab(in)`` acquires the element that has been received during an ``onPush()``. It cannot be called again until the
+  port is pushed again by the upstream.
+* ``cancel(in)`` closes the input port.
+
+The events corresponding to an *input* port can be received in an :class:`InHandler` instance registered to the
+input port using ``setHandler(in, handler)``. This handler has three callbacks:
+
+* ``onPush()`` is called when the output port has now a new element. Now it is possible to aquire this element using
+  ``grab()`` and/or call ``pull(in)`` on the port to request the next element. It is not mandatory to grab the
+  element, but if it is pulled while the element has not been grabbed it will drop the buffered element.
+* ``onUpstreamFinish()`` is called once the upstream has completed and no longer can be pulled for new elements.
+  No more ``onPush()`` will arrive after this event. If not overridden this will default to stopping the stage.
+* ``onUpstreamFailure()`` is called if the upstream failed with an exception and no longer can be pulled for new elements.
+  No more ``onPush()`` will arrive after this event. If not overridden this will default to failing the stage.
+
+Also, there are three query methods available for input ports:
+
+* ``isAvailable(out)`` returns true if the port can be grabbed.
+* ``hasBeenPulled(out)`` returns true if the port has been already pulled. Calling ``pull(in)`` in this state is illegal.
+* ``isClosed`` returns true if the port is closed. At this point the port can not be pulled and will not be pushed anymore.
+
+The relationship of the above operations, events and queries are summarized in the state machine below. Green shows
+the initial state while orange indicates the end state. If an operation is not listed for a state, then it is invalid
+to call it while the port is in that state. If an event is not listed for a state, then that event cannot happen
+in that state.
+
+|
+
+.. image:: ../images/inport_transitions.png
+  :align: center
+
+|
+
+Finally, there are two methods available for convenience to complete the stage and all of its ports:
+
+* ``completeStage()`` is equivalent to closing all output ports and cancelling all input ports.
+* ``failStage(exception)`` is equivalent to failing all output ports and cancelling all input ports.
+
+
+Using timers
+------------
+
+**TODO**
+
+Using asynchronous side-channels
+--------------------------------
+
+**TODO**
+
+Integration with actors
+-----------------------
+
+**TODO**
+
+Built-in shapes
+---------------
+
+**TODO**
+
+Custom materialized values
+--------------------------
+
+**TODO**
+
+Using attriubtes to affect the behavior of a stage
+--------------------------------------------------
+
+**TODO**
+
+
+Custom linear processing stages with PushPullStage
+==================================================
+
+.. warning::
```

Comment:
I will remove this. Probably too early to deprecate, let's have feedback form the GraphStage first.